observatory-platform/observatory/platform/elastic/elastic.py

```diff
@@ -0,0 +1,187 @@
+    """ Make an Elasticsearch URI.
+
+    :param schema: the schema, e.g. http or https.
+    :param user: Elasticsearch username.
+    :param secret: Elasticsearch secret.
+    :param hostname: Elasticsearch hostname.
+    :param port: Elasticsearch port.
+    :return: the full Elasticsearch URI.
+    """
+
+    return f"{schema}://{user}:{secret}@{hostname}:{port}"
+
+
+def make_index_prefix(feed_name: str, table_name: str):
+    """ Make an index prefix from a feed name and a table name.
+
+    :param feed_name: the feed name.
+    :param table_name: the table name
+    :return: the index prefix.
+    """
+
+    return f"{feed_name}-{table_name}"
+
+
+class Elastic:
+    def __init__(
+        self,
+        host: str = "http://elasticsearch:9200/",
+        thread_count: int = cpu_count(),
+        chunk_size: int = 10000,
+        timeout: int = 30000,
+    ):
+        """ Create an Elastic API client.
+
+        :param host: the host including the hostname and port.
+        :param thread_count: the number of threads to use when loading data into Elastic.
+        :param chunk_size: the batch size for loading documents.
+        :param timeout: the timeout in seconds.
+        """
+
+        self.host = host
+        self.thread_count = thread_count
+        self.chunk_size = chunk_size
+        self.timeout = timeout
+        self.es = Elasticsearch(hosts=[self.host], timeout=timeout)
+
+        # logging.basicConfig()
+        # logging.getLogger().setLevel(logging.WARNING)
+
+    def delete_index(self, index_id: str) -> None:
+        """ Delete an Elastic index.
+
+        :param index_id: the index ID.
+        :return: None.
+        """
+
+        try:
+            self.es.indices.delete(index=index_id)
+        except elasticsearch.exceptions.TransportError as e:
+            logging.warning(f"Index not found: {e}")
+
+    def get_alias_indexes(self, alias_id: str) -> List:
+        """ Return the list of indexes associated with an alias.
+
+        :param alias_id: the alias identifier.
+        :return: the list of indexes.
+        """
+
+        index_ids = []
+        try:
+            indexes: Dict = self.es.indices.get(alias_id)
+            for key, val in indexes.items():
+                index_ids.append(key)
+        except elasticsearch.exceptions.NotFoundError:
+            pass
+
+        return index_ids
+
+    def index_documents(self, index_id: str, mappings: Dict, documents: Iterator[Dict]) -> bool:
+        """ Index documents supplied as an iterator.
+
+        :param index_id: the index ID.
+        :param mappings: the mappings supplied as a Dict.
+        :param documents: the documents supplied as an iterator.
+        :return: whether the documents were loaded successfully or not.
+        """
+
+        # Load mapping and create index
+
+        body = {
+            "settings": {
+                "index": {"number_of_shards": 1, "number_of_replicas": 1},
```

Comment:
this section is what I'm not sure about, it probably needs to be put into environment variables or some sort of config. But perhaps for now, set shards to 3 and leave replicates to 1